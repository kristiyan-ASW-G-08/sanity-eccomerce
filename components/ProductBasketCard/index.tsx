import { FC } from 'react';
import { useDispatch } from 'react-redux';
import {
  addProduct,
  BasketProduct,
  removeProduct,
  removeProductFromBasket,
} from '@/slices/BasketSlice';
import ImageUrlBuilder from '@sanity/image-url';
import client from 'client';
const ProductBasketCard: FC<BasketProduct> = ({
  name,
  _id,
  price,
  currentVariant,
  quantity,
}) => {
  const dispatch = useDispatch();

  return (
    <article className="flex  w-full">
      <div className="flex flex-col md:flex-row md:max-w-xl rounded-lg bg-white shadow-lg  w-full">
        <div className="bg-neutral-300">
          <img
            src={ImageUrlBuilder(client)
              .image(currentVariant?.image)
              .width(250)
              .url()}
            alt="Sunset in the mountains"
          />
        </div>

        <div className="p-6  grid space-y-3">
          <h5 className="text-2xl font-medium mb-2">{name}</h5>

          <div className="space-x-4 flex">
            <button
              onClick={() => {
                dispatch(removeProduct(`${_id}-${currentVariant.variantName}`));
              }}
              className="bg-red-400 text-neutral-50 text-3xl rounded-full w-10 h-10 grid place-center"
            >
              -
            </button>

            <button
              onClick={() => {
                dispatch(
                  addProduct({
                    name,
                    _id,
                    price,
                    currentVariant,
                    quantity,
                    basketIdentifier: `${_id}-${currentVariant.variantName}`,
                  }),
                );
              }}
              className="bg-blue-400 text-neutral-50 text-3xl rounded-full w-10 h-10 grid place-center"
            >
              +
            </button>
            <p className="w-28 h-10">Quantity {quantity}</p>
            <p className="text-gray-600 text-1xl">${price}</p>
          </div>
          <button
            className="bg-red-400 hover:bg-red-700 text-neutral-50"
            onClick={() => {
              dispatch(
                removeProductFromBasket(`${_id}-${currentVariant.variantName}`),
              );
            }}
          >
            Remove
          </button>
        </div>
      </div>
    </article>
  );
};

export default ProductBasketCard;
