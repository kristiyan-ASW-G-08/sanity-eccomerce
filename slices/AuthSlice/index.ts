import {
  createSlice,
  PayloadAction,
  createEntityAdapter,
} from '@reduxjs/toolkit';
import { PURGE } from 'redux-persist';
export interface AuthState {
  _id: string;
  token: string;
  email: string;
  isAdmin: boolean;
}

export const initialState = {
  token: '',
  email: '',
  _id: '',
  isAdmin: false,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    login: (state: any, action: PayloadAction<AuthState>) => {
      return action.payload;
    },
    logout: (state: any, action: PayloadAction<AuthState>) => {
      return initialState;
    },
  },
  extraReducers: builder => {
    builder.addCase(PURGE, state => {
      return initialState;
    });
  },
});

export const { login, logout } = authSlice.actions;
export default authSlice.reducer;
