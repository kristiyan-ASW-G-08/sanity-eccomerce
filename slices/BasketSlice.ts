import {
  createSlice,
  PayloadAction,
  createEntityAdapter,
} from '@reduxjs/toolkit';
import { PURGE } from 'redux-persist';
export interface BasketProduct {
  _id: string;
  basketIdentifier: string;
  name: string;
  price: number;
  quantity: number;
  currentVariant: {
    image: { asset: { ref: string } };
    variantName: string;
    color: string;
  };
}
type _id = string;
const initialState: BasketProduct[] = [];

const basketSlice = createSlice({
  name: 'basket',
  initialState,
  reducers: {
    addProduct: (
      state: BasketProduct[],
      action: PayloadAction<BasketProduct>,
    ) => {
      const hasProduct = state.find(
        product => product.basketIdentifier == action.payload.basketIdentifier,
      );
      if (hasProduct) {
        const updatedState = state.map(product => {
          if (product.basketIdentifier === action.payload.basketIdentifier) {
            return { ...product, quantity: product.quantity + 1 };
          }
          return product;
        });
        return updatedState;
      } else {
        return [...state, action.payload];
      }
    },
    removeProduct: (state: BasketProduct[], action: PayloadAction<string>) => {
      console.log(action.payload);
      const product = state.find(
        product => product.basketIdentifier === action.payload,
      );
      if (product && product.quantity === 1) {
        return [
          ...state.filter(
            product => product.basketIdentifier !== action.payload,
          ),
        ];
      } else if (product) {
        return [
          ...state.filter(
            product => product.basketIdentifier !== action.payload,
          ),
          { ...product, quantity: product.quantity - 1 },
        ];
      }
    },
    removeProductFromBasket: (
      state: BasketProduct[],
      action: PayloadAction<string>,
    ) => {
      const product = state.find(
        product => product.basketIdentifier === action.payload,
      );

      return [
        ...state.filter(product => product.basketIdentifier !== action.payload),
      ];
    },
    resetBasket: (state: BasketProduct[], action: PayloadAction) => {
      return initialState;
    },
  },
  extraReducers: builder => {
    builder.addCase(PURGE, state => {
      return initialState;
    });
  },
});

export const { addProduct, removeProduct, resetBasket,removeProductFromBasket } = basketSlice.actions;
export default basketSlice.reducer;
