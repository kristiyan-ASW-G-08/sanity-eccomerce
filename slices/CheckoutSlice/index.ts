import {
  createSlice,
  PayloadAction,
  createEntityAdapter,
} from '@reduxjs/toolkit';
import { PURGE } from 'redux-persist';
export interface Checkout {
  address: string;
  phoneNumber: string;
  zip: string;
  city: string;
  fullName: string;
  country: string;
}

export const initialState = {
  address: '',
  phoneNumber: '',
  zip: '',
  city: '',
  fullName: '',
  country: '',
};

const authSlice = createSlice({
  name: 'checkout',
  initialState,
  reducers: {
    registerCheckout: (state: any, action: PayloadAction<Checkout>) => {
      return action.payload;
    },
    removeCheckout: (state: any, action: PayloadAction) => {
      return initialState;
    },
  },
  extraReducers: builder => {
    builder.addCase(PURGE, state => {
      return initialState;
    });
  },
});

export const { registerCheckout, removeCheckout } = authSlice.actions;
export default authSlice.reducer;
