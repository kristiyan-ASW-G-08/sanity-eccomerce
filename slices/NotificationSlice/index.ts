import {
  createSlice,
  PayloadAction,
  createEntityAdapter,
} from '@reduxjs/toolkit';
import { PURGE } from 'redux-persist';
export interface Notification {
  content: string;
  type: 'message' | 'alert';
  isActive: boolean;
}

export const initialState = {
  content: '',
  type: 'alert',
  isActive: false,
};

const notificationSlice = createSlice({
  name: 'checkout',
  initialState,
  reducers: {
    setNotification: (state: any, action: PayloadAction<Notification>) => {
      return action.payload;
    },
    removeNotification: (state: any, action: PayloadAction<Notification>) => {
      return initialState;
    },
  },
  extraReducers: builder => {
    builder.addCase(PURGE, state => {
      return initialState;
    });
  },
});

export const { setNotification, removeNotification } =
  notificationSlice.actions;
export default notificationSlice.reducer;
