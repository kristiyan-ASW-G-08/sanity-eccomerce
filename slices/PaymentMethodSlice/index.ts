import {
  createSlice,
  PayloadAction,
  createEntityAdapter,
} from '@reduxjs/toolkit';
import { PURGE } from 'redux-persist';

export type method = 'Paypal' | 'Debit or Credit Card' | '';
export interface PaymentMethod {
  method: method;
}

export const initialState = {
  method: '',
};

const paymentReducer = createSlice({
  name: 'payment',
  initialState,
  reducers: {
    registerPaymentMethod: (
      state: any,
      action: PayloadAction<PaymentMethod>,
    ) => {
      return action.payload;
    },
    removePaymentMethod: (state: any, action: PayloadAction) => {
      return initialState;
    },
  },
  extraReducers: builder => {
    builder.addCase(PURGE, state => {
      return initialState;
    });
  },
});

export const { registerPaymentMethod, removePaymentMethod } =
  paymentReducer.actions;
export default paymentReducer.reducer;
