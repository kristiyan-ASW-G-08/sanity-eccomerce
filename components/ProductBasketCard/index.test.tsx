import React from 'react';
import { render, waitFor } from '@testing-library/react';
import ProductBasketCard from '.';
import UserEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import Store from '@/store/index';
import { removeProduct, addProduct } from '@/slices/BasketSlice';

jest.mock('@/slices/BasketSlice');

describe('ProductBasketCard', () => {
  afterEach(() => jest.resetAllMocks());
  afterAll(() => jest.restoreAllMocks());
  it('render ', async () => {
    const product = {
      name: 'Product',
      _id: '1',
      price: 1000,
      image: 'someimage',
      quantity: 3,
      stock: 10,
    };
    const { container, getByText } = render(<ProductBasketCard {...product} />, {
      wrapper: ({ children }) => <Provider store={Store}>{children}</Provider>,
    });

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();

    const addButton = getByText('Add');
    const removeButton = getByText('Remove');
    UserEvent.click(addButton);

    waitFor(() => {
      expect(addProduct).toHaveBeenCalledTimes(1);
    });



    UserEvent.click(removeButton);

    waitFor(() => {
      expect(removeProduct).toHaveBeenCalledTimes(1);
    });

  });
});
