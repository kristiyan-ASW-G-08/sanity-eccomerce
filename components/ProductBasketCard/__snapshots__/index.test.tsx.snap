// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBasketCard render  1`] = `
<div>
  <article
    class="flex  w-full"
  >
    <div
      class="flex flex-col md:flex-row md:max-w-xl rounded-lg bg-white shadow-lg  w-full"
    >
      <img
        alt=""
        class=" h-24 w-24 object-contain"
      />
      <div
        class="p-6 flex flex-col justify-start space-y-3"
      >
        <h5
          class="text-2xl font-medium mb-2"
        >
          Product
        </h5>
        <p
          class="text-gray-600 text-1xl"
        >
          $
          1000
        </p>
        <div
          class="space-x-4 flex"
        >
          <button
            class="bg-red-400 text-neutral-50 text-1xl p-1"
          >
            Remove
          </button>
          <button
            class="bg-blue-400 text-neutral-50 text-1xl  p-1"
          >
            Add
          </button>
          <p>
            3
          </p>
        </div>
      </div>
    </div>
  </article>
</div>
`;
